### PR TITLE
[merge] instantiate and upgrade

### LINKED ISSUE #2655: [v2 merge] As a user I can instantiate and upgrade v1 chaincodes
## Description

Need to be able to follow the two types of chain code lifecycle. Currently v1 extension follows v1 lifecycle, and v2 extension follows v2 lifecycle.

- [ ] Probably need to implement fabric-client (1.4) functionality
- [ ] Re-introduce instantiate command (this lifecycle step does not exist in the v2 lifecycle)

### PR DESCRIPTION
**Pull request by lquintai**
Closes #2655 

Signed-off-by: Leonor Quintais <lquintai@uk.ibm.com>

### CODE REVIEW

File: packages/blockchain-environment-v1/src/FabricEnvironmentConnection.ts [hunk 1]
```diff
@@ -500,4 +498,30 @@ export class FabricEnvironmentConnection implements IFabricEnvironmentConnection
 
         return peer.getAllChannelNames();
     }
+
+    private async instantiateOrUpgradeChaincode(_name: string, _version: string, _peerNames: Array<string>, _channelName: string, _fcn: string, _args: Array<string>, _collectionPath: string, _contractEP: any, isUpgrade: boolean): Promise<any> { // FIXME was buffer
```
Comment (Jakeeyturner) [inline packages/blockchain-environment-v1/src/FabricEnvironmentConnection.ts:502]: `instantiateOrUpgradeSmartContractDefinition` seems to return `void` (is this correct?), so we should make this `Promise<void>`?

Same for `instantiateChaincode`

File: packages/blockchain-environment-v1/test/FabricEnvironmentConnection.test.ts [hunk 2]
```diff
@@ -1247,15 +1247,51 @@ describe('FabricEnvironmentConnection', () => {
 
     describe('instantiateChaincode', () => {
         it('should instantiate the specified chaincode', async () => {
-            const payload: Buffer = await connection.instantiateChaincode('myChaincode', '0.0.1', ['peer0.org1.example.com', 'peer0.org2.example.com'], 'mychannel', 'instantiate', ['arg1'], path.join('myPath'), undefined);
-            payload.should.deep.equal(Buffer.from('TODO'));
+            const collectionDef: FabricCollectionDefinition[] = '{ some json content }' as unknown as FabricCollectionDefinition[];
```
Comment (Jakeeyturner) [inline packages/blockchain-environment-v1/test/FabricEnvironmentConnection.test.ts:1250]: Could we make this an actual collection definition?
```new FabricCollectionDefinition('myCollection', `OR('Org1MSP)`, 4, 2);``` for example

File: packages/blockchain-extension/extension/commands/UserInputUtil.ts [hunk 3]
```diff
@@ -1279,16 +1279,18 @@ export class UserInputUtil {
...
+    private static async getInstalledContracts(connection: IFabricEnvironmentConnection, peers: Array<string>, isV1: boolean = false): Promise<IBlockchainQuickPickItem<{ packageEntry: PackageRegistryEntry, workspace: vscode.WorkspaceFolder }>[]> {
         const tempQuickPickItems: IBlockchainQuickPickItem<{ packageEntry: PackageRegistryEntry, workspace: vscode.WorkspaceFolder }>[] = [];
         for (const peer of peers) {
-            const chaincodes: FabricInstalledSmartContract[] = await connection.getInstalledSmartContracts(peer);
+            const chaincodes: FabricInstalledSmartContract[] = await connection.getInstalledSmartContracts(peer, isV1);
 
             // TODO: this is wrong but won't be needed later
             chaincodes.forEach((chaincode: FabricInstalledSmartContract) => {
-                const _package: PackageRegistryEntry = new PackageRegistryEntry({ name: chaincode.label, version: chaincode.packageId, path: undefined, sizeKB: undefined });
+                const nameAndVersion: string[] = chaincode.label.split('@');
```
Comment (Jakeeyturner) [inline packages/blockchain-extension/extension/commands/UserInputUtil.ts:1290]: I _think_ in our v1 extension we allow a user to instantiate a contract with a package such as `packageWithoutVersion.cds` - might be worth double-checking that and confirming that this code will work okay if that's the case.

File: packages/blockchain-extension/extension/commands/instantiateCommand.ts [hunk 4]
```diff
@@ -111,7 +110,9 @@ export async function instantiateSmartContract(channelName?: string, peerNames?:
 
             if (doInstall) {
                 // Install smart contract package
-                packageEntry = await vscode.commands.executeCommand(ExtensionCommands.INSTALL_SMART_CONTRACT, undefined, peerNames, packageToInstall);
+                const channelMap: Map<string, string[]> = await connection.createChannelMap();
+                packageEntry = await vscode.commands.executeCommand(ExtensionCommands.INSTALL_SMART_CONTRACT, channelMap, packageToInstall);
+                // TODO: at the moment successful install of v1 contract doesn't return packageEntry, so we will never progress after this point.
```
Comment (Jakeeyturner) [inline packages/blockchain-extension/extension/commands/instantiateCommand.ts:115]: Is this still the case?

File: packages/blockchain-fabric-admin/src/LifecycleChannel.ts [hunk 5]
```diff
@@ -124,6 +126,49 @@ export class LifecycleChannel {
...
+        if (!args) {
+            args = [];
+        }
+        for (const arg of args) {
+            specArgs.push(Buffer.from(arg, 'utf8'));
+        }
+
+        const ccSpec: any = {
+            // TODO Golang is defaul in v1 fabric-client, and v1 exension always used default. Should we try to extract language from packaged contracts somehow?
+            type: protos.protos.ChaincodeSpec.Type.GOLANG,
```
Comment (Jakeeyturner) [inline packages/blockchain-fabric-admin/src/LifecycleChannel.ts:147]: Any idea how the type is being used? 👀

File: packages/blockchain-extension/extension/commands/instantiateCommand.ts (hunk 4 above)
Comment (lquintai) [reply packages/blockchain-extension/extension/commands/instantiateCommand.ts:115]: yes

File: packages/blockchain-fabric-admin/src/LifecycleChannel.ts (hunk 5 above)
Comment (lquintai) [reply packages/blockchain-fabric-admin/src/LifecycleChannel.ts:147]: Had a quick look, only see spec being used for name and version. Will need to look at it better later on.
https://github.com/hyperledger/fabric/blob/cae9a63b5dd54adef10e4a27a93e24a87e675f6f/core/scc/lscc/lscc.go#L807

**Review by Jakeeyturner — APPROVED** [review]

**Pull request merged**